### PR TITLE
[Qt] Add support for Table widget

### PR DESCRIPTION
**Pull request by corranwebster**
This implements the Table widget for the Qt backend. No support for embedded widgets, but otherwise complete. Should be reasonably performant, but if there are issues, we can probably refine further (eg. by setting constant row height).

Qt is particularly finicky about warning about widgets in calls, so we need a warning check around anything which might cause Qt to call the data() method of the cells, which means a slight change in the tests but which shouldn't affect outcomes.

This doesn't implement user-adjustable column widths: rather it stretches the columns evenly into the available space. I could get the columns to start even and then be user-resizable, but not pass the tests - I think I was setting things up too early; column size policies probably need to be set when the widget is parented is my guess.

Other than that, fairly straightforward, and the data() method logic is particularly clean following #4042 being merged.

Ref #3914

To Do:

- [x] screenshot
- [ ] Open an issue to get interactive re-sizing working.

## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by freakboy3742** [comment]
As a heads up - I wouldn't be *overly* concerned about adding support for widgets in table cells. They're only supported on macOS, and even there, I'm not convinced we've got the API right. They're flagged as a beta feature in the docs; I wouldn't consider them a requirement for landing this PR.

**Comment by corranwebster** [comment]
> As a heads up - I wouldn't be _overly_ concerned about adding support for widgets in table cells. They're only supported on macOS, and even there, I'm not convinced we've got the API right. They're flagged as a beta feature in the docs; I wouldn't consider them a requirement for landing this PR.

I wasn't planning to implement the widgets, the thing that was being difficult was the tests for the warnings about not handling them.

Currently the blocker is getting the columns roughly evenly spaced - code which is working fine when in examples is failing in the tests (plus KDE and MacOS versions of Qt having different behaviours for column widths!)

**Comment by corranwebster** [comment]
> Currently the blocker is getting the columns roughly evenly spaced - code which is working fine when in examples is failing in the tests (plus KDE and MacOS versions of Qt having different behaviours for column widths!)

@freakboy3742  Question: how important is for the user to be able to resize the columns? I can get the tests to pass perfectly if I use `ResizeMode.Stretch` (evenly sized columns, but no interactive resizing) but if I use `ResizeMode.Interactive` it wants to use minimum widths for the columns and let the user re-size.

**Comment by freakboy3742** [comment]
> @freakboy3742 Question: how important is for the user to be able to resize the columns? I can get the tests to pass perfectly if I use `ResizeMode.Stretch` (evenly sized columns, but no interactive resizing) but if I use `ResizeMode.Interactive` it wants to use minimum widths for the columns and let the user re-size.

Manual resizing *should* be possible... but I'll also accept that perfect is the enemy of the good. If you want to put up a PR with a "non-resizable" columns, and log the resize issue as a bug to be resolved later, I'd be OK with that.

**Comment by corranwebster** [comment]
Ready for review. iOS test failure should be unrelated.

File: testbed/tests/widgets/test_table.py [hunk 1]
```diff
@@ -612,14 +612,14 @@ async def test_cell_widget(widget, probe):
...
-
-    if probe.supports_widgets:
-        probe.assert_cell_content(0, 2, widget=widget.data[0].c)
-        probe.assert_cell_content(1, 2, widget=widget.data[1].c)
-    else:
-        # If the platform doesn't support cell widgets, the test should still *run* -
-        # we just won't have widgets in the cells.
-        probe.assert_cell_content(0, 2, "MISSING!")
-        probe.assert_cell_content(1, 2, "MISSING!")
+        probe.assert_cell_content(1, 0, "A1")
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_table.py]: What's the thinking with this change? `pytest.warns` asserts that the warning occurred; the original intention was to make that test as narrow as possible so that we can confirm that a specific action yields a specific warning.

If the issue here is that the subsequent actions also raise warnings that are noisy, then I'd argue there should be a separate `with warning_check` context manager.

File: qt/src/toga_qt/widgets/table.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        return len(self._columns)
+
+    def data(
+        self,
+        index: QModelIndex | QPersistentModelIndex,
+        /,
+        role: int = Qt.ItemDataRole.DisplayRole,
+    ) -> Any:
+        if not index.isValid():
+            return None  # pragma: no cover
```
Comment (freakboy3742) [inline qt/src/toga_qt/widgets/table.py]: `#pragma` should always be a last resort; and when it's used, we try to document *why* it has been used.

From the look of it, most of these fall into the category of "safety catch for an edge case that *should* never happen, but just in case...". If I've understood that correctly, we should add an inline comment to that effect at the top of the method. If it's not, and there's a realistic possibility where these edge cases could be triggered, we should add cases to the testbed to exercise them, even if they're no-ops (or at least uncontroversial) for other backends.

Comment (corranwebster) [reply qt/src/toga_qt/widgets/table.py]: Yes, most of these are "this should never happen but just in case..." mainly because we are dependent on the Python side telling us when changes to the data happens, and if it doesn't then we could end up with weirdness. We could add tests for misbehaving data sources generally to check robustness, but I feared going down that path and ending up having to fix a bunch of other backends.  I'll open an issue on this.

There are a few which are in fact needed (like the checks for `source` being `None` which can occur briefly during initialisation but don't seem to be picked up in the tests).

Finally, looking at this, I probably should put a try-except around a bunch of this to guard against user error like bugs in custom data sources.

It would be nice to be able to run unit tests on complex Qt subclasses like `TableSourceModel` independently - is there a way to run backend-specific unit tests outside the testbed?

File: testbed/tests/widgets/test_table.py (hunk 1 above)
Comment (corranwebster) [reply testbed/tests/widgets/test_table.py]: The problem here is that on Qt, the probe asking for *any* data will trigger the warning, because unlike some backends we aren't storing any representation of the data inside the model, it's all queried lazily.

Thinking about this, I guess I could have the probe suppress the warnings internally, so we only see the warnings generated by the widget doing its stuff.

File: qt/src/toga_qt/widgets/table.py (hunk 2 above)
Comment (corranwebster) [reply qt/src/toga_qt/widgets/table.py]: I've added issue #4101 to note that we should test custom data sources (and decide what to do if they are bad).

Comment (freakboy3742) [reply qt/src/toga_qt/widgets/table.py]: Right now, there's no mechanism for "platform-specific runtime tests". 

In *theory*, there shouldn't be a need for any - if there's an API edge case that can cause an error on Qt, there's no reason to believe the same edge case wouldn't cause problems on any other platform. And even if it doesn't, having a few more tests verifying normal operation doesn't hurt.

Of course, that then means that for cases like `TableSourceModel`, you don't write a "unit" test - you have to use the public API that is exercising the table model, which I appreciate can be hard to manufacture.

I'm not fundamentally opposed to adding platform-specific runtime tests, though. My guess is that the best approach would be to come up with some sort of convention for a folder in `tests_backend` (e.g., `tests_backend/tests`) that is included in the test discovery process of testbed. Most backends won't provide one; but if Qt needs to add some tests to exercise those edge cases, then we can.

I wouldn't consider that a pre-requisite for merging this PR, though.

**Comment by freakboy3742** [comment]
See #4109 for the enhancement request around column resizing.
